### PR TITLE
Added typed construtors for operators FockBasis

### PR DESCRIPTION
**Pull request by myrddin89**
I initiated to apply the changes as discussed in #24. Before completing I think we should review the changes I made so far. I think there are some options also for the docs. The changes I made in `fock.jl` for the operators are what I am used to. I can modify, if you consider to be better another way.

### CODE REVIEW

File: src/fock.jl [hunk 1]
```diff
@@ -21,44 +21,63 @@ end
...
-    destroy(b::FockBasis)
+    destroy([T=ComplexF64,] b::FockBasis)
 
-Annihilation operator for the specified Fock space.
+Annihilation operator for the specified Fock space with optional data type `T`.
 """
-function destroy(b::FockBasis)
-    diag = complex.(sqrt.(b.offset+1.:b.N))
+function destroy(::Type{C}, b::FockBasis) where C
+    T = real(C)
```
Comment (david-pl) [inline src/fock.jl:43]: Could you explain why you're taking the detour over `real(C)` here?

Comment (myrddin89) [reply src/fock.jl:43]: Sorry I was a bit busy today. I always struggle with this issue. One major concern I usually have is to get the right type in computations, especially when using extended precision datatypes (the reason why I have proposed these changes in the first place). 

Assuming that one specifies the datatype as `Complex{T}` for some `T` we still needs to compute the `sqrt`s with the correct precision, as by default `sqrt` uses `Float64` if the input type is not an `AbstractFloat`.

Another path we could follow, is to always assume the specified type is real and construct the state/operator with the correct complex type.

Comment (david-pl) [reply src/fock.jl:43]: Thanks for the explanation, that makes sense.

**Comment by david-pl** [comment]
The changes so far look good to me!

**Pull request merged**